sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/package.scala

```diff
@@ -86,11 +88,42 @@ package object expressions  {
   /**
    * Helper functions for working with `Seq[Attribute]`.
    */
-  implicit class AttributeSeq(attrs: Seq[Attribute]) {
+  implicit class AttributeSeq(val attrs: Seq[Attribute]) {
     /** Creates a StructType with a schema matching this `Seq[Attribute]`. */
     def toStructType: StructType = {
       StructType(attrs.map(a => StructField(a.name, a.dataType, a.nullable)))
     }
+
+    // It's possible that `attrs` is a linked list, which can lead to bad O(n^2) loops when
+    // accessing attributes by their ordinals. To avoid this performance penalty, convert the input
+    // to an array.
+    private lazy val attrsArray = attrs.toArray
+
+    private lazy val exprIdToOrdinal = {
+      val arr = attrsArray
+      val map = Maps.newHashMapWithExpectedSize[ExprId, Int](arr.length)
+      var index = 0
+      while (index < arr.length) {
+        val exprId = arr(index).exprId
+        if (!map.containsKey(exprId)) {
+          map.put(exprId, index)
```

Comment:
I was being conservative here in order to match the behavior of the old linear scan, which stopped upon finding the first entry. However, we can remove the need for this check if we iterate over `arr` in reverse-order.